src/pyk/proof/reachability.py

```diff
@@ -141,44 +207,112 @@ def advance_proof(
                 self.proof.kcfg.create_edge(curr_node.id, next_node.id, depth)
+                self.proof.kcfg.add_expanded(curr_node.id)
                 _LOGGER.info(
                     f'Found basic block at depth {depth} for {self.proof.id}: {shorten_hashes((curr_node.id, next_node.id))}.'
                 )
-                curr_node = next_node
-
-                if self._check_subsumption(
-                    kcfg_explore, curr_node, is_terminal=is_terminal, implication_every_block=implication_every_block
-                ):
-                    continue
+                continue
 
+            # Stuck
             if len(next_cterms) == 0:
+                self.proof.kcfg.add_expanded(curr_node.id)
                 _LOGGER.info(f'Found stuck node {self.proof.id}: {shorten_hashes(curr_node.id)}')
+                continue
+
+            # Cut Rule
+            if len(next_cterms) == 1:
+                next_node = self.proof.kcfg.get_or_create_node(next_cterms[0])
+                self.proof.kcfg.create_edge(curr_node.id, next_node.id, 1)
+                self.proof.kcfg.add_expanded(curr_node.id)
+                _LOGGER.info(
+                    f'Inserted cut-rule basic block at depth 1 for {self.proof.id}: {shorten_hashes((curr_node.id, next_node.id))}.'
+                )
+                continue
 
-            elif len(next_cterms) > 1:
+            # Branching
+            if len(next_cterms) > 1:
                 branches = list(extract_branches(cterm)) if extract_branches is not None else []
                 if len(branches) != len(next_cterms):
                     _LOGGER.warning(
                         f'Falling back to manual branch extraction {self.proof.id}: {shorten_hashes(curr_node.id)}'
                     )
-                    branches = [mlAnd(c for c in s.constraints if c not in cterm.constraints) for s in next_cterms]
+                    # TODO: should not have to prune bottom branches, the backend should do this for us.
+                    _next_cterms = []
+                    for ct in next_cterms:
+                        _LOGGER.info(f'Checking for bottom branch: {ct}')
+                        _ct = kcfg_explore.cterm_simplify(ct)
+                        if is_bottom(_ct):
+                            _LOGGER.warning(f'Found bottom branch: {ct}')
+                        else:
+                            _next_cterms.append(CTerm.from_kast(_ct))
+                    branches = [mlAnd(c for c in s.constraints if c not in cterm.constraints) for s in _next_cterms]
                 _LOGGER.info(
                     f'Found {len(branches)} branches for node {self.proof.id}: {shorten_hashes(curr_node.id)}: {[kcfg_explore.kprint.pretty_print(bc) for bc in branches]}'
                 )
                 self.proof.kcfg.split_on_constraints(curr_node.id, branches)
+                continue
+
+        self.proof.write_proof()
+        return self.proof.kcfg
+
+
+class AGBMCProver(AGProver):
```

Comment:
If `AGBMCProof` is no longer an `AGProof`, then  you'll probably need this change as well:

```suggestion
class AGBMCProver(Prover):
```

(Also, consider changing the name to `BMCProver`.)

Reply:
Here is the proposed new design: https://github.com/runtimeverification/pyk/issues/361